source/user/lmbench3/doc/lat_udp.8
.\" $Id: lat_udp.8,v 1.1.1.1 2006-11-23 11:25:34 steven Exp $
.TH LAT_UDP 8 "$Date: 2006-11-23 11:25:34 $" "(c)1994 Larry McVoy" "LMBENCH"
.SH NAME
lat_udp \- measure interprocess communication latency via UDP/IP
.SH SYNOPSIS
.B lat_udp
.I -s
.sp .5
.B lat_udp
[
.I "-m <message size>"
]
[
.I "-P <parallelism>"
]
[
.I "-W <warmups>"
]
[
.I "-N <repetitions>"
]
.I hostname
.sp .5
.B lat_udp
.I "-S hostname"
.SH DESCRIPTION
.B lat_udp
is a client/server program that measures interprocess
communication latencies.  The benchmark passes a message back and forth between
the two processes (this sort of benchmark is frequently referred to as a
``hot potato'' benchmark).  No other work is done in the processes.
.LP
.B lat_udp
has three forms of usage: as a server (-s), as a client (lat_udp localhost), and
as a shutdown (lat_udp -S localhost).
.SH OUTPUT
The reported time is in microseconds per round trip and includes the total
time, i.e., the context switching overhead is included.
Output format is like so
.sp
.ft CB
UDP latency using localhost: 650 microseconds
.ft
.SH ACKNOWLEDGEMENT
Funding for the development of
this tool was provided by Sun Microsystems Computer Corporation.
.SH "SEE ALSO"
lmbench(8), lat_fcntl(8), lat_fifo(8), lat_tcp(8), lat_unix(8).
.SH "AUTHOR"
Carl Staelin and Larry McVoy
.PP
Comments, suggestions, and bug reports are always welcome.


source/History
Ralink AP SDK - History

Version 4.0.1.0
===============

- [IPv6] Avoid both eth2.2 and eth2 set same IPV6 address after receiving RA packet in gateway mode.
- [IPv6] Fix HNAT misbinding issue (RT6855/6)
- [IPv6] Fix ping fail(ping ::1) issue of loopback interface when we try to enable and disable disable_ipv6 parameter in /proc directory (RT6855/6)
- [Switch] Fix legacy switch(RT3052/RT3352/RT5350) settings vlan 12345 error
- [PPPoE] Kernel panic after PAP fail (RT6855/6)
- [PPTP] Kernel panic after PPTP peer terminated the link (RT6855/6)
- [SPI Flash] There are some noise if we try to program flash while I2S is playing music (RT3052/RT3883/RT3352/RT5350)
- [GPIO] Fix GPIO interrupt handler cannot be trigged correctly (RT6855/6)
- [GUI] Fix WPS bug for test plan
- [NET] Recover irqs_disabled() check under SKB Recycle to fix warning messages, but it decreases RT6855/6's maximum peek throughput
- [Uboot] RT6855/6 does watchdog reset at cold start to fix timer interrupt issue

Version 4.0.0.0
===============

PLATFORM:

- support RT3052/RT3883/RT3352/RT5350 (linux-2.6.21.x)
- support RT6855/RT6856 (linux-2.6.36MT.x)

NEW FUNCTIONS:

- [IPv6 ready logo Phase II] RT6855/RT6856
- [IPv6] support IPv6 Rapid Deployment
- [IPv6] support IPv6 DS-Lite
- [GUI] support two giga phy port
- [GUI] display AP Client site_survey

IMPROVEMENT:

- [IPv6] support ipv6 tftp server
- [PPTP/L2TP] performace tunning at Linux 2.6.21 (PPTP improve 25%, L2TP improve 10% compare to SDK 3.6)
- [SAMBA] significant performance fine tune on full loading scenario (dual-band concurrent AP with USB access)
- [NAND Flash] config page size and address cycle based on chip mode
- [I2S] Add MMAP for user/kernel
- [GUI] support WPS 2.0 

Bug fixed:

- [Kernel NVRAM] memory leakage / initiate failed / free cache / wirte data exceed 1024
- [RAETH] fix STP drop issue on RT3052 chipst ONLY
- [RAETH] fix switch abnormal after ifconfig down&up
- [RAETH] fix PDMA hang issue under heavy load test
- [RAETH] fix TSO bug if USB storage is enabled
- [USB device] system will hang whiling loading gadget driver. Turning off power-saving on both port0 and port1
- [USB device] force host to do re-enumeration to solve "unrecognize device" when re-plug-in in windows platform.
- [IPv6] kernel panic bug if enable ipv6 privacy extension or router prereference
- [IPv6] fix NS packet loss for ipv6 DAD
- [GUI] flush conntrack table while restart/deleting rules
- [NAND Flash] fix nand flash not work for MTD_ANY mode
- [IGMPPROXY] fixed the overflow FATAL bug while handling bit47=1 mac address
- [GPIO] fixed RT3352/RT5350 extended GPIO pins mapping



source/user/lmbench3/src/lat_udp.c
/*
 * udp_xact.c - simple UDP transaction latency test
 *
 * Three programs in one -
 *	server usage:	lat_udp -s
 *	client usage:	lat_udp [-P <parallelism>] [-W <warmup>] [-N <repetitions>] hostname
 *	shutdown:	lat_udp -S hostname
 *
 * Copyright (c) 1994 Larry McVoy.  Distributed under the FSF GPL with
 * additional restriction that results may published only if
 * (1) the benchmark is unmodified, and
 * (2) the version in the sccsid below is included in the report.
 * Support for this development by Sun Microsystems is gratefully acknowledged.
 */
char	*id = "$Id: lat_udp.c,v 1.1.1.1 2006-11-23 11:25:34 steven Exp $\n";
#include "bench.h"

#define MAX_MSIZE (10 * 1024 * 1024)

void	client_main(int ac, char **av);
void	server_main();
void	timeout();
void	init(iter_t iterations, void* cookie);
void	cleanup(iter_t iterations, void* cookie);
void    doit(iter_t iterations, void* cookie);

typedef struct _state {
	int	sock;
	int	seq;
	int	msize;
	char	*server;
	char	*buf;
} state_t;


int
main(int ac, char **av)
{
	state_t state;
	int	c;
	int	parallel = 1;
	int	warmup = 0;
	int	repetitions = TRIES;
	int	server = 0;
	int	shutdown = 0;
	int	msize = 4;
 	char	buf[256];
	char	*usage = "-s\n OR [-S] [-m <message size>] [-P <parallelism>] [-W <warmup>] [-N <repetitions>] server\n NOTE: message size must be >= 4\n";

	if (sizeof(int) != 4) {
		fprintf(stderr, "lat_udp: Wrong sequence size\n");
		return(1);
	}

	while (( c = getopt(ac, av, "sS:m:P:W:N:")) != EOF) {
		switch(c) {
		case 's': /* Server */
			if (fork() == 0) {
				server_main();
			}
			exit(0);
		case 'S': /* shutdown serverhost */
		{
			int seq, n;
			int sock = udp_connect(optarg,
					       UDP_XACT,
					       SOCKOPT_NONE);
			for (n = -1; n > -5; --n) {
				seq = htonl(n);
				(void) send(sock, &seq, sizeof(int), 0);
			}
			close(sock);
			exit (0);
		}
		case 'm':
			msize = atoi(optarg);
			if (msize < sizeof(int)) {
				lmbench_usage(ac, av, usage);
				msize = 4;
			}
			if (msize > MAX_MSIZE) {
				lmbench_usage(ac, av, usage);
				msize = MAX_MSIZE;
			}
			break;
		case 'P':
			parallel = atoi(optarg);
			if (parallel <= 0)
				lmbench_usage(ac, av, usage);
			break;
		case 'W':
			warmup = atoi(optarg);
			break;
		case 'N':
			repetitions = atoi(optarg);
			break;
		default:
			lmbench_usage(ac, av, usage);
			break;
		}
	}

	if (optind + 1 != ac) {
		lmbench_usage(ac, av, usage);
	}

	state.server = av[optind];
	state.msize = msize;
	benchmp(init, doit, cleanup, SHORT, parallel, 
		warmup, repetitions, &state);
	sprintf(buf, "UDP latency using %s", state.server);
	micro(buf, get_n());
	exit(0);
}

void
init(iter_t iterations, void* cookie)
{
	state_t *state = (state_t *) cookie;

	if (iterations) return;

	state->sock = udp_connect(state->server, UDP_XACT, SOCKOPT_NONE);
	state->seq = 0;
	state->buf = (char*)malloc(state->msize);
	
	signal(SIGALRM, timeout);
	alarm(15);
}

void
doit(iter_t iterations, void *cookie)
{
	state_t *state = (state_t *) cookie;
	int seq = state->seq;
	int net = htonl(seq);
	int sock = state->sock;
	int ret;

	alarm(15);
	while (iterations-- > 0) {
		*(int*)state->buf = htonl(seq++);
		if (send(sock, state->buf, state->msize, 0) != state->msize) {
			perror("lat_udp client: send failed");
			exit(5);
		}
		if (recv(sock, state->buf, state->msize, 0) != state->msize) {
			perror("lat_udp client: recv failed");
			exit(5);
		}
	}
	state->seq = seq;
}

void
cleanup(iter_t iterations, void* cookie)
{
	state_t *state = (state_t *) cookie;

	if (iterations) return;

	close(state->sock);
	free(state->buf);
}

void
timeout()
{
	fprintf(stderr, "Recv timed out\n");
	exit(1);
}

void
server_main()
{
	char	*buf = (char*)valloc(MAX_MSIZE);
	int     sock, sent, namelen, seq = 0;
	struct sockaddr_in it;

	GO_AWAY;

	sock = udp_server(UDP_XACT, SOCKOPT_REUSE);

	while (1) {
		int nbytes;
		namelen = sizeof(it);
		if ((nbytes = recvfrom(sock, (void*)buf, MAX_MSIZE, 0, 
		    (struct sockaddr*)&it, &namelen)) < 0) {
			fprintf(stderr, "lat_udp server: recvfrom: got wrong size\n");
			exit(9);
		}
		sent = ntohl(*(int*)buf);
		if (sent < 0) {
			udp_done(UDP_XACT);
			exit(0);
		}
		if (sent != ++seq) {
			seq = sent;
		}
		*(int*)buf = htonl(seq);
		if (sendto(sock, (void*)buf, nbytes, 0, 
		    (struct sockaddr*)&it, sizeof(it)) < 0) {
			perror("lat_udp sendto");
			exit(9);
		}
	}
}
